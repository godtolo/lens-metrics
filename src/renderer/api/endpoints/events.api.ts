import moment from "moment";
import { KubeObject } from "../kube-object";
import { formatDuration } from "../../utils/formatDuration";
import { autoBind } from "../../utils";
import { KubeApi } from "../kube-api";
import { KubeJsonApiData } from "../kube-json-api";

export class KubeEvent extends KubeObject {
  static kind = "Event";
  static namespaced = true;
  static apiBase = "/api/v1/events";

  constructor(data: KubeJsonApiData) {
    super(data);
    autoBind(this);
    Object.assign(this, data); // TODO: figure out why this doesn't work from super()
  }

  involvedObject: {
    kind: string;
    namespace: string;
    name: string;
    uid: string;
    apiVersion: string;
    resourceVersion: string;
    fieldPath: string;
  };
  reason: string;
  message: string;
  source: {
    component: string;
    host: string;
  };
  firstTimestamp: string;
  lastTimestamp: string;
  count: number;
  type: "Normal" | "Warning" | string;
  eventTime: null;
  reportingComponent: string;
  reportingInstance: string;

  isWarning() {
    return this.type === "Warning";
  }

  getSource() {
    const { component, host } = this.source;

    return `${component} ${host || ""}`;
  }

  getFirstSeenTime() {
    const diff = moment().diff(this.firstTimestamp);

    return formatDuration(diff, true);
  }

  getLastSeenTime() {
    const diff = moment().diff(this.lastTimestamp);

    return formatDuration(diff, true);
  }
}

export const eventApi = new KubeApi({
  objectConstructor: KubeEvent,
});
