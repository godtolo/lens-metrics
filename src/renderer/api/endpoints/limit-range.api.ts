import { KubeObject } from "../kube-object";
import { KubeApi } from "../kube-api";
import { autoBind } from "../../utils";
import { KubeJsonApiData } from "../kube-json-api";

export enum LimitType {
  CONTAINER = "Container",
  POD = "Pod",
  PVC = "PersistentVolumeClaim",
}

export enum Resource {
  MEMORY = "memory",
  CPU = "cpu",
  STORAGE = "storage",
  EPHEMERAL_STORAGE = "ephemeral-storage",
}

export enum LimitPart {
  MAX = "max",
  MIN = "min",
  DEFAULT = "default",
  DEFAULT_REQUEST = "defaultRequest",
  MAX_LIMIT_REQUEST_RATIO = "maxLimitRequestRatio",
}

type LimitRangeParts = Partial<Record<LimitPart, Record<string, string>>>;

export interface LimitRangeItem extends LimitRangeParts {
  type: string
}

export class LimitRange extends KubeObject {
  static kind = "LimitRange";
  static namespaced = true;
  static apiBase = "/api/v1/limitranges";

  constructor(data: KubeJsonApiData) {
    super(data);
    autoBind(this);
  }

  spec: {
    limits: LimitRangeItem[];
  };

  getContainerLimits() {
    return this.spec.limits.filter(limit => limit.type === LimitType.CONTAINER);
  }

  getPodLimits() {
    return this.spec.limits.filter(limit => limit.type === LimitType.POD);
  }

  getPVCLimits() {
    return this.spec.limits.filter(limit => limit.type === LimitType.PVC);
  }
}

export const limitRangeApi = new KubeApi({
  objectConstructor: LimitRange,
});
