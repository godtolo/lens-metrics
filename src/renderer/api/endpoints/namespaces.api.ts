import { KubeApi } from "../kube-api";
import { KubeObject } from "../kube-object";
import { autoBind } from "../../utils";
import { KubeJsonApiData } from "../kube-json-api";

export enum NamespaceStatus {
  ACTIVE = "Active",
  TERMINATING = "Terminating",
}

export class Namespace extends KubeObject {
  static kind = "Namespace";
  static namespaced = false;
  static apiBase = "/api/v1/namespaces";

  constructor(data: KubeJsonApiData) {
    super(data);
    autoBind(this);
  }

  status?: {
    phase: string;
  };

  getStatus() {
    return this.status ? this.status.phase : "-";
  }
}

export const namespacesApi = new KubeApi({
  objectConstructor: Namespace,
});
