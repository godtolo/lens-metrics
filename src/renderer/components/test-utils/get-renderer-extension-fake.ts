/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { SetRequired } from "type-fest";
import { LensRendererExtension } from "../../../extensions/lens-renderer-extension";

export class TestExtension extends LensRendererExtension {}

export const getRendererExtensionFake = ({ id, ...rest }: SetRequired<Partial<LensRendererExtension>, "id">) => {
  const instance = new TestExtension({
    id,
    absolutePath: "irrelevant",
    isBundled: false,
    isCompatible: false,
    isEnabled: false,
    manifest: { name: id, version: "some-version" },
    manifestPath: "irrelevant",
  });

  Object.assign(instance, rest);

  return instance;
};
