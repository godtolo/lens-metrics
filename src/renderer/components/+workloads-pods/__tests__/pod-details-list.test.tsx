/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { AsyncFnMock } from "@async-fn/jest";
import asyncFn from "@async-fn/jest";
import type { DiContainer } from "@ogre-tools/injectable";
import { fireEvent } from "@testing-library/react";
import React from "react";
import directoryForLensLocalStorageInjectable from "../../../../common/directory-for-lens-local-storage/directory-for-lens-local-storage.injectable";
import fetchInjectable, { Fetch } from "../../../../common/fetch/fetch.injectable";
import { Pod } from "../../../../common/k8s-api/endpoints";
import { getDiForUnitTesting } from "../../../getDiForUnitTesting";
import storesAndApisCanBeCreatedInjectable from "../../../stores-apis-can-be-created.injectable";
import { DiRender, renderFor } from "../../test-utils/renderFor";
import { PodDetailsList } from "../pod-details-list";
import type { PodStore } from "../store";
import podStoreInjectable from "../store.injectable";

const runningPod = new Pod({
  apiVersion: "foo",
  kind: "Pod",
  metadata: {
    name: "foobar",
    resourceVersion: "foobar",
    uid: "foobar",
    ownerReferences: [{
      uid: "runningReplicaSet",
      apiVersion: "v1",
      kind: "ReplicaSet",
      name: "running",
    }],
    namespace: "default",
    selfLink: "/apis/apps/v1/replicasets/default/foobar",
  },
  status: {
    phase: "Running",
    conditions: [
      {
        type: "Initialized",
        status: "True",
        lastProbeTime: 1,
        lastTransitionTime: "1",
      },
      {
        type: "Ready",
        status: "True",
        lastProbeTime: 1,
        lastTransitionTime: "1",
      },
    ],
    hostIP: "10.0.0.1",
    podIP: "10.0.0.1",
    startTime: "now",
    containerStatuses: [],
    initContainerStatuses: [],
  },
});

const pendingPod = new Pod({
  apiVersion: "foo",
  kind: "Pod",
  metadata: {
    name: "foobar-pending",
    resourceVersion: "foobar",
    uid: "foobar-pending",
    ownerReferences: [{
      uid: "pendingReplicaSet",
      apiVersion: "v1",
      kind: "ReplicaSet",
      name: "pending",
    }],
    namespace: "default",
    selfLink: "/apis/apps/v1/replicasets/default/foobar-pending",
  },
});

const failedPod = new Pod({
  apiVersion: "foo",
  kind: "Pod",
  metadata: {
    name: "foobar-failed",
    resourceVersion: "foobar",
    uid: "foobar-failed",
    ownerReferences: [{
      uid: "failedReplicaSet",
      apiVersion: "v1",
      kind: "ReplicaSet",
      name: "failed",
    }],
    namespace: "kube-system",
    selfLink: "/apis/apps/v1/replicasets/default/foobar-failed",
  },
  status: {
    phase: "Failed",
    conditions: [],
    hostIP: "10.0.0.1",
    podIP: "10.0.0.1",
    startTime: "now",
  },
});

describe("<PodDetailsList />", () => {
  let di: DiContainer;
  let podStore: PodStore;
  let render: DiRender;
  let fetchMock: AsyncFnMock<Fetch>;

  beforeEach(() => {
    di = getDiForUnitTesting({ doGeneralOverrides: true });
    di.override(storesAndApisCanBeCreatedInjectable, () => true);
    di.override(
      directoryForLensLocalStorageInjectable,
      () => "/some-directory-for-lens-local-storage",
    );

    fetchMock = asyncFn();
    podStore = di.inject(podStoreInjectable);

    di.override(fetchInjectable, () => fetchMock);

    render = renderFor(di);
  });
  
  describe("when no pods passed and podStore is loaded", () => {
    beforeEach(() => {
      podStore.isLoaded = true;
    });

    it("renders", () => {
      const result = render(
        <PodDetailsList
          pods={[]}
        />
      );

      expect(result.container).toMatchSnapshot();
    });

    it("shows empty message", () => {
      const result = render(
        <PodDetailsList
          pods={[]}
        />
      );

      expect(result.getByText("No items found")).toBeInTheDocument();
    });
  });

  describe("when podStore is still loading", () => {
    it("renders spinner", () => {
      const result = render(
        <PodDetailsList
          pods={[]}
        />
      );

      expect(result.container).toMatchSnapshot();
    });
  });

  describe("when few pods passed", () => {
    beforeEach(() => {
      podStore.isLoaded = true;
    });

    it("renders table with pods", () => {
      const result = render(
        <PodDetailsList
          pods={[
            runningPod,
            failedPod,
            pendingPod,
          ]}
        />
      );

      expect(result.container).toMatchSnapshot();
    });

    describe("when metrics max values passed", () => {
      beforeEach(() => {
        podStore.getPodKubeMetrics = (pod: Pod) => {
          let cpu = 0.4;
          let memory = 1024;

          switch(pod.getId()) {
            case "foobar":
              cpu = 0.2;
              memory = 128;
              break;
            case "foobar-pending":
              cpu = 0.5;
              memory = 256;
          }

          return {
            cpu,
            memory
          }
        };
      });

      it("renders table with metric lines", () => {
        const result = render(
          <PodDetailsList
            pods={[
              runningPod,
              failedPod,
              pendingPod,
            ]}
            maxCpu={1}
            maxMemory={2048}
          />
        );

        expect(result.container).toMatchSnapshot();
      });

      it("sorts table by name", () => {
        const result = render(
          <PodDetailsList
            pods={[
              runningPod,
              failedPod,
              pendingPod,
            ]}
            maxCpu={1}
            maxMemory={2048}
          />
        );

        const column = result.getByText("Name");

        fireEvent.click(column);
        fireEvent.click(column);

        expect(result.container).toMatchSnapshot();
      });

      it("sorts table by cpu", () => {
        const result = render(
          <PodDetailsList
            pods={[
              runningPod,
              failedPod,
              pendingPod,
            ]}
            maxCpu={1}
            maxMemory={2048}
          />
        );

        const column = result.getByText("CPU");

        fireEvent.click(column);
        fireEvent.click(column);

        expect(result.container).toMatchSnapshot();
      });

      it("sorts table by memory", () => {
        const result = render(
          <PodDetailsList
            pods={[
              runningPod,
              failedPod,
              pendingPod,
            ]}
            maxCpu={1}
            maxMemory={2048}
          />
        );

        const column = result.getByText("Memory");

        fireEvent.click(column);
        fireEvent.click(column);

        expect(result.container).toMatchSnapshot();
      });
    });
  });
});