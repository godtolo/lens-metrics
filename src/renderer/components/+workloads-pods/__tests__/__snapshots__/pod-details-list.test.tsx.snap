// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PodDetailsList /> when few pods passed renders table with pods 1`] = `
<div>
  <div
    class="PodDetailsList flex column"
  >
    <div
      class="DrawerTitle title"
    >
      Pods
    </div>
    <div
      class="Table flex column box grow selectable sortable autoSize"
    >
      <div
        class="TableHead sticky"
      >
        <div
          class="TableCell name sorting"
        >
          Name
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node sorting"
        >
          Node
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell namespace sorting"
        >
          Namespace
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell ready"
        >
          Ready
        </div>
        <div
          class="TableCell cpu sorting"
        >
          CPU
          <i
            class="Icon sortIcon enabled material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell memory sorting"
        >
          Memory
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell status"
        >
          Status
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          default
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          0
        </div>
        <div
          class="TableCell memory"
        >
          0
        </div>
        <div
          class="TableCell status running"
        >
          Running
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar-failed
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          kube-system
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          0
        </div>
        <div
          class="TableCell memory"
        >
          0
        </div>
        <div
          class="TableCell status failed"
        >
          Failed
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar-pending
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          default
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          0
        </div>
        <div
          class="TableCell memory"
        >
          0
        </div>
        <div
          class="TableCell status waiting"
        >
          Waiting
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PodDetailsList /> when few pods passed when metrics max values passed renders table with metric lines 1`] = `
<div>
  <div
    class="PodDetailsList flex column"
  >
    <div
      class="DrawerTitle title"
    >
      Pods
    </div>
    <div
      class="Table flex column box grow selectable sortable autoSize"
    >
      <div
        class="TableHead sticky"
      >
        <div
          class="TableCell name sorting"
        >
          Name
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node sorting"
        >
          Node
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell namespace sorting"
        >
          Namespace
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell ready"
        >
          Ready
        </div>
        <div
          class="TableCell cpu sorting"
        >
          CPU
          <i
            class="Icon sortIcon enabled material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell memory sorting"
        >
          Memory
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell status"
        >
          Status
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar-pending
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          default
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 50.00%;"
            />
          </div>
          <div>
            <p>
              CPU: 50%
              <br />
              0.500
            </p>
          </div>
        </div>
        <div
          class="TableCell memory"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 12.50%;"
            />
          </div>
          <div>
            <p>
              Memory: 13%
              <br />
              256.000B
            </p>
          </div>
        </div>
        <div
          class="TableCell status waiting"
        >
          Waiting
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar-failed
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          kube-system
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 40.00%;"
            />
          </div>
          <div>
            <p>
              CPU: 40%
              <br />
              0.400
            </p>
          </div>
        </div>
        <div
          class="TableCell memory"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 50.00%;"
            />
          </div>
          <div>
            <p>
              Memory: 50%
              <br />
              1.000KiB
            </p>
          </div>
        </div>
        <div
          class="TableCell status failed"
        >
          Failed
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          default
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 20.00%;"
            />
          </div>
          <div>
            <p>
              CPU: 20%
              <br />
              0.200
            </p>
          </div>
        </div>
        <div
          class="TableCell memory"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 6.25%;"
            />
          </div>
          <div>
            <p>
              Memory: 7%
              <br />
              128.000B
            </p>
          </div>
        </div>
        <div
          class="TableCell status running"
        >
          Running
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PodDetailsList /> when few pods passed when metrics max values passed sorts table by cpu 1`] = `
<div>
  <div
    class="PodDetailsList flex column"
  >
    <div
      class="DrawerTitle title"
    >
      Pods
    </div>
    <div
      class="Table flex column box grow selectable sortable autoSize"
    >
      <div
        class="TableHead sticky"
      >
        <div
          class="TableCell name sorting"
        >
          Name
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node sorting"
        >
          Node
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell namespace sorting"
        >
          Namespace
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell ready"
        >
          Ready
        </div>
        <div
          class="TableCell cpu sorting"
        >
          CPU
          <i
            class="Icon sortIcon enabled material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell memory sorting"
        >
          Memory
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell status"
        >
          Status
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar-pending
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          default
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 50.00%;"
            />
          </div>
          <div>
            <p>
              CPU: 50%
              <br />
              0.500
            </p>
          </div>
        </div>
        <div
          class="TableCell memory"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 12.50%;"
            />
          </div>
          <div>
            <p>
              Memory: 13%
              <br />
              256.000B
            </p>
          </div>
        </div>
        <div
          class="TableCell status waiting"
        >
          Waiting
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar-failed
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          kube-system
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 40.00%;"
            />
          </div>
          <div>
            <p>
              CPU: 40%
              <br />
              0.400
            </p>
          </div>
        </div>
        <div
          class="TableCell memory"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 50.00%;"
            />
          </div>
          <div>
            <p>
              Memory: 50%
              <br />
              1.000KiB
            </p>
          </div>
        </div>
        <div
          class="TableCell status failed"
        >
          Failed
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          default
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 20.00%;"
            />
          </div>
          <div>
            <p>
              CPU: 20%
              <br />
              0.200
            </p>
          </div>
        </div>
        <div
          class="TableCell memory"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 6.25%;"
            />
          </div>
          <div>
            <p>
              Memory: 7%
              <br />
              128.000B
            </p>
          </div>
        </div>
        <div
          class="TableCell status running"
        >
          Running
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PodDetailsList /> when few pods passed when metrics max values passed sorts table by memory 1`] = `
<div>
  <div
    class="PodDetailsList flex column"
  >
    <div
      class="DrawerTitle title"
    >
      Pods
    </div>
    <div
      class="Table flex column box grow selectable sortable autoSize"
    >
      <div
        class="TableHead sticky"
      >
        <div
          class="TableCell name sorting"
        >
          Name
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node sorting"
        >
          Node
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell namespace sorting"
        >
          Namespace
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell ready"
        >
          Ready
        </div>
        <div
          class="TableCell cpu sorting"
        >
          CPU
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell memory sorting"
        >
          Memory
          <i
            class="Icon sortIcon enabled material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell status"
        >
          Status
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar-failed
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          kube-system
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 40.00%;"
            />
          </div>
          <div>
            <p>
              CPU: 40%
              <br />
              0.400
            </p>
          </div>
        </div>
        <div
          class="TableCell memory"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 50.00%;"
            />
          </div>
          <div>
            <p>
              Memory: 50%
              <br />
              1.000KiB
            </p>
          </div>
        </div>
        <div
          class="TableCell status failed"
        >
          Failed
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar-pending
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          default
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 50.00%;"
            />
          </div>
          <div>
            <p>
              CPU: 50%
              <br />
              0.500
            </p>
          </div>
        </div>
        <div
          class="TableCell memory"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 12.50%;"
            />
          </div>
          <div>
            <p>
              Memory: 13%
              <br />
              256.000B
            </p>
          </div>
        </div>
        <div
          class="TableCell status waiting"
        >
          Waiting
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          default
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 20.00%;"
            />
          </div>
          <div>
            <p>
              CPU: 20%
              <br />
              0.200
            </p>
          </div>
        </div>
        <div
          class="TableCell memory"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 6.25%;"
            />
          </div>
          <div>
            <p>
              Memory: 7%
              <br />
              128.000B
            </p>
          </div>
        </div>
        <div
          class="TableCell status running"
        >
          Running
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PodDetailsList /> when few pods passed when metrics max values passed sorts table by name 1`] = `
<div>
  <div
    class="PodDetailsList flex column"
  >
    <div
      class="DrawerTitle title"
    >
      Pods
    </div>
    <div
      class="Table flex column box grow selectable sortable autoSize"
    >
      <div
        class="TableHead sticky"
      >
        <div
          class="TableCell name sorting"
        >
          Name
          <i
            class="Icon sortIcon enabled material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node sorting"
        >
          Node
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell namespace sorting"
        >
          Namespace
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell ready"
        >
          Ready
        </div>
        <div
          class="TableCell cpu sorting"
        >
          CPU
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell memory sorting"
        >
          Memory
          <i
            class="Icon sortIcon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </div>
        <div
          class="TableCell status"
        >
          Status
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar-pending
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          default
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 50.00%;"
            />
          </div>
          <div>
            <p>
              CPU: 50%
              <br />
              0.500
            </p>
          </div>
        </div>
        <div
          class="TableCell memory"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 12.50%;"
            />
          </div>
          <div>
            <p>
              Memory: 13%
              <br />
              256.000B
            </p>
          </div>
        </div>
        <div
          class="TableCell status waiting"
        >
          Waiting
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar-failed
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          kube-system
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 40.00%;"
            />
          </div>
          <div>
            <p>
              CPU: 40%
              <br />
              0.400
            </p>
          </div>
        </div>
        <div
          class="TableCell memory"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 50.00%;"
            />
          </div>
          <div>
            <p>
              Memory: 50%
              <br />
              1.000KiB
            </p>
          </div>
        </div>
        <div
          class="TableCell status failed"
        >
          Failed
        </div>
      </div>
      <div
        class="TableRow nowrap"
      >
        <div
          class="TableCell name"
        >
          foobar
        </div>
        <div
          class="TableCell warning"
        />
        <div
          class="TableCell node"
        />
        <div
          class="TableCell namespace"
        >
          default
        </div>
        <div
          class="TableCell ready"
        >
          0 / 0
        </div>
        <div
          class="TableCell cpu"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 20.00%;"
            />
          </div>
          <div>
            <p>
              CPU: 20%
              <br />
              0.200
            </p>
          </div>
        </div>
        <div
          class="TableCell memory"
        >
          <div
            class="LineProgress"
          >
            <div
              class="line"
              style="width: 6.25%;"
            />
          </div>
          <div>
            <p>
              Memory: 7%
              <br />
              128.000B
            </p>
          </div>
        </div>
        <div
          class="TableCell status running"
        >
          Running
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PodDetailsList /> when no pods passed and podStore is loaded renders 1`] = `
<div>
  <div>
    No items found
  </div>
</div>
`;

exports[`<PodDetailsList /> when podStore is still loading renders spinner 1`] = `
<div>
  <div
    class="PodDetailsList flex justify-center"
  >
    <div
      class="Spinner singleColor"
    />
  </div>
</div>
`;
